crowdsourcehinter/crowdsourcehinter.py

```diff
@@ -0,0 +1,352 @@
+        Specifically this function is used to send necessary data to JS about incorrect answer
+        submissions and hints. It also will return hints that have been reported, although this
+        is only for Staff.
+        Returns:
+          used_hint_answer_text: This dicitonary contains reported hints/answers (if the user is staff) and the
+                         first hint/answer pair that the student submitted for a problem.
+        """
+        # used_hint_answer_text is a dictionary of hints (or lack thereof) used for a
+        # specific answer, as well as 2 other random hints that exist for each answer
+        # that were not used. The keys are the used hints, the values are the
+        # corresponding incorrect answer
+        used_hint_answer_text = {}
+        if self.get_user_is_staff():
+            for key in self.reported_hints:
+                used_hint_answer_text[key] = str("Reported")
+        if len(self.incorrect_answers) == 0:
+            return used_hint_answer_text
+        else:
+            for index in range(0, len(self.used)):
+                # each index is a hint that was used, in order of usage
+                if str(self.used[index]) in self.hint_database[self.incorrect_answers[index]]:
+                    # add new key (hint) to used_hint_answer_text with a value (incorrect answer)
+                    used_hint_answer_text[str(self.used[index])] = str(self.incorrect_answers[index])
+                    self.incorrect_answers = []
+                    self.used = []
+                    return used_hint_answer_text
+                else:
+                    # if the student's answer had no hints (or all the hints were reported and unavailable) return None
+                    used_hint_answer_text[None] = str(self.incorrect_answers[index])
+                    self.incorrect_answers = []
+                    self.used = []
+                    return used_hint_answer_text
+        self.incorrect_answers = []
+        self.used = []
+        return used_hint_answer_text
+
+    @XBlock.json_handler
+    def rate_hint(self, data, suffix=''):
+        """
+        Used to facilitate hint rating by students.
+        Hint ratings in hint_database are updated and the resulting hint rating (or reported status) is returned to JS.
+        Args:
+          data['student_answer']: The incorrect answer that corresponds to the hint that is being rated
+          data['hint']: The hint that is being rated
+          data['student_rating']: The rating chosen by the student.
+        Returns:
+          'rating': the new rating of the hint, or the string 'reported' if the hint was reported
+          'hint': the hint that had its rating changed
+        """
+        answer_data = data['student_answer']
+        data_rating = data['student_rating']
+        data_hint = data['hint']
+        if data_hint == 'Sorry, there are no hints for this answer.':
```

Comment:
Minor: This is fragile way to communicate between browser and server, and will definitely not work if we ever localize this code.